### DISCUSSION TITLE
ATtiny841 Pin mapping issue

### DISCUSSION THREAD

**Discussion by CaptainCheeze29** [SpenceKonde/ATTinyCore#837] [Support / Q & A]
Hi there! First time poster and also very basic coder and designer here.
Possibly someone could point out how I'm going wrong here. Maybe it's something obvious I've missed.

I've been working on a USB programmable ATtiny841 based dev board as a hobby/learning project to see if my self-teaching could actually make something like this work. And to some extent I've actually prototyped and got it working perfectly!....almost. 
I'm just having an odd issue with the pin mapping when writing a fairly simple Arduino IDE code controlling an WS2812B led matrix. 
My dev board schematic is shown below:
[ATtiny841 Dev Board with UART.pdf](https://github.com/SpenceKonde/ATTinyCore/files/13989722/ATtiny841.Dev.Board.with.UART.pdf)

Using the ATtinyCore and Arduino as ISP, I flashed the ATtiny441/841 (Optibootloader) with the following settings clearly using the Clockwise pin mapping:

![image](https://github.com/SpenceKonde/ATTinyCore/assets/94358321/b5e9aa5b-fa99-4721-a19a-4be07228887c)

Now everything works find and dandy with being able to upload code to the Attiny841, no issue there...however, when uploading my code I am experiencing a weird issue. 
I have two switches controlling the light functions of my leds: they are connected to the arduino pins 2 and 3 (as per the ATtinyCore pinout graphics) and are written in the code as such. They work perfectly. Great!
My issue comes with the controlling of the LED matrix. I have it connected to Pin13 of the Attiny (arduino pin 0 in clockwise mapping), but when writing it as Arduino pin 0 in the code it then displays the output on pin2 of the Attiny (arduino pin 10 in the mapping). And the reverse is true if I program it to Arduino pin 10, the output displays on Arduino pin0 (currently how I have it in the code provided below). Basically, it seems that the LED matrix output gets programmed to the CCW pin mapping while the switches use the regular CW pin mapping.

Have I got a hardware issue or is there something I'm missing here?? I know I can get the code to work and all, but it would be nice if the pin that I code for was the actual pin that gets used. 
Thanks in advance.

```
//ATtiny841 FastLED RGB Switch Control

#include <FastLED.h>

#define ledPin 10
#define numLed 25
#define Bright 50
#define Order GRB
#define chipType WS2812B
#define Volts 5
#define Amps 500
#define dt 100

CRGB leds[numLed];

#define colourButt 2
int colourVal;
#define brightButt 3
int brightVal;

uint8_t colourChng;
uint8_t brightChng;
uint8_t hue = 0;

void fadingColorPattern(int delayTime) {
  static int hue = 0;

  for (int i = 0; i < numLed; i++) {
    leds[i] = CHSV(hue, 255, 255);
  }

  FastLED.show();
  delay(delayTime);

  hue += 2;  // Adjust the increment to change the speed of color change
  if (hue >= 256) {
    hue = 0;  // Reset hue after it reaches the maximum value
  }
}

#define FLOW_SPEED    100         // Adjust the flow speed
void matrixCodeEffect() {
  static uint8_t codeRain[numLed];
  
  // Move the existing code down
  for (int i = numLed - 1; i > 0; i--) {
    codeRain[i] = codeRain[i - 1];
  }
  
  // Add a new "code" at the top
  codeRain[0] = random8();

  // Display the matrix code
  for (int i = 0; i < numLed; i++) {
    leds[i] = CRGB::Green;  // Color of the "code"
    leds[i].fadeToBlackBy(codeRain[i]);  // Fade the code as it moves down
  }

  FastLED.show();
  delay(FLOW_SPEED);
}

void setup() {
  // put your setup code here, to run once:
  //Serial.begin(9600);         //Only required for testing on Arduino UNO
  pinMode(colourButt, INPUT);
  digitalWrite(colourButt, HIGH);
  pinMode(brightButt, INPUT);
  digitalWrite(brightButt, HIGH);

  FastLED.addLeds<chipType, ledPin, Order>(leds, numLed);
  FastLED.setMaxPowerInVoltsAndMilliamps(Volts, Amps);
  FastLED.setBrightness(Bright);
  FastLED.setCorrection(TypicalPixelString);
  FastLED.clear();
  FastLED.show();
}

void loop() {
  // put your main code here, to run repeatedly:
  colourVal = digitalRead(colourButt);
  if (colourVal == 1) {

  } else if (colourVal == 0) {
    colourChng++;
    if (colourChng > 11) colourChng = 0;
    delay(500);
  }
  switch (colourChng) {
    case 0:
      fill_solid(leds, numLed, CRGB::Red); //Red
      break;
    case 1:
      fill_solid(leds, numLed, CRGB::Blue); //Blue
      break;
    case 2:
      fill_solid(leds, numLed, CRGB::Green); //Green
      break;
    case 3:
      fill_solid(leds, numLed, CRGB(250, 0, 255)); //Pink
      break;
    case 4:
      fill_solid(leds, numLed, CRGB(80, 0, 150)); //Royal Purple
      break;
    case 5:
      fill_solid(leds, numLed, CRGB(43, 149, 43)); //Light Green
      break;
    case 6:
      fill_solid(leds, numLed, CRGB(65, 65, 255)); //Light Blue
      break;
    case 7:
      fill_solid(leds, numLed, CRGB(255, 128, 0)); //Orange
      break;
    case 8:
      fill_solid(leds, numLed, CRGB(218, 165, 32)); //Goldenrod
      break;
    case 9:
      static uint8_t startIndex = 0;

  // Fill the LEDs with the RainbowColors_p palette
  for (int i = 0; i < numLed; i++) {
    leds[i] = ColorFromPalette(RainbowColors_p, startIndex + i * (256 / numLed), 255);
  }
  FastLED.show();
  // Change the start index for the next iteration
  startIndex += 2;  // Adjust the increment to change the speed of color change
  delay(20);  // Adjust the delay for the speed of the rainbow effect
      break;
    case 10:
      fadingColorPattern(75);  // Adjust the delay for fading speed
      break;
    case 11:
      matrixCodeEffect();
      break;
  }
  brightVal = digitalRead(brightButt);
  if (brightVal == 1) {

  } else if (brightVal == 0) {
    brightChng++;
    if (brightChng > 4) brightChng = 0;
    delay(500);
  }
  switch (brightChng) {
    case 0:
      FastLED.setBrightness(40);
      break;
    case 1:
      FastLED.setBrightness(30);
      break;
    case 2:
      FastLED.setBrightness(20);
      break;
    case 3:
      FastLED.setBrightness(10);
      break;
    case 4:
      FastLED.setBrightness(5);
      break;
  }
  FastLED.show();
}
```

**Comment by SpenceKonde** [comment]
Aaaha! That sheds some light on this issue....

**Reply by CaptainCheeze29** [reply]
Ok, so this is a known issue then and I'm not going mad? XD

**Comment by ahshah** [comment]
Could you simplify the code? If you can simplify, others can recreate the scenario and offer advice. Pin addressing problems could probably be expressed in ~20 lines of code, and this simplicity makes it easier to dive in to the different dimensions of pin issues. (i.e pin mapping selected, the method of addressing pins).

**Comment by CaptainCheeze29** [comment]
Sorry it's taken me ages to get back to this. Here is another simplified code that does a similar thing. For some reason, regardless of the pin mapping option I choose, it seems to use the counterclockwise pin mapping on the dev board.
But just lately, now I'm having issues using the usb serial communication completely. Board is not taking the code. I've tried to redo the bootloader on the Attiny841, but it just will not take the code. It's happened to two chips in the last couple days that were both working. I'll make a seperate post for that issue though with the verbose info.

```
#include <FastLED.h>

#define DATA_PIN 4 //pin 2 on attiny84
#define LED_TYPE WS2812B
#define COLOR_ORDER GRB
#define NUM_LEDS 90
CRGB leds[NUM_LEDS];

#define BRIGHTNESS 96
#define FRAMES_PER_SECOND 255

uint8_t gHue = 0; // rotating "base color" used by many of the patterns

void setup() {
  delay(3000);  // 3 second delay for recovery

  // tell FastLED about the LED strip configuration
  FastLED.addLeds<LED_TYPE, DATA_PIN, COLOR_ORDER>(leds, NUM_LEDS).setCorrection(TypicalLEDStrip);
  //FastLED.addLeds<LED_TYPE,DATA_PIN,CLK_PIN,COLOR_ORDER>(leds, NUM_LEDS).setCorrection(TypicalLEDStrip);

  // set master brightness control
  FastLED.setBrightness(BRIGHTNESS);
}

void loop() {
  // Call the current pattern function once, updating the 'leds' array
  rainbow();

  // send the 'leds' array out to the actual LED strip
  FastLED.show();
  // insert a delay to keep the framerate modest
  FastLED.delay(1000 / FRAMES_PER_SECOND);
  EVERY_N_MILLISECONDS( 20 ) { gHue++; } // slowly cycle the "base color" through the rainbow
}

void rainbow() 
{
  // FastLED's built-in rainbow generator
  fill_rainbow( leds, NUM_LEDS, gHue, 7);
}
```
